projects/i18n/src/lib/i18n.module.ts

```diff
@@ -0,0 +1,85 @@
+/*!
+ * Copyright 2020 VMware, Inc.
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+import { TranslationLoader } from './loader/translation-loader';
+import { ModuleWithProviders, Optional, Provider } from '@angular/core';
+import { NgModule, InjectionToken, Inject } from '@angular/core';
+import { FormatDateTimePipe } from './pipe/format-date-time-pipe';
+import { MessageFormatTranslationService } from './service/message-format-translation-service';
+import { TranslationService } from './service/translation-service';
+import { BOOTSTRAP_DETAILS } from './utils/tokens';
+import { TranslationPipe } from './pipe/translation-pipe';
+import { HttpClient } from '@angular/common/http';
+
+let singletonService: TranslationService = null;
+
+export function genericSingletonFactory(details: { locale: string }): TranslationService {
+    if (singletonService === null) {
+        singletonService = new MessageFormatTranslationService(details.locale, 'en');
+    }
+    return singletonService;
+}
+
+/**
+ * A module that mananges translation capabilites for the application.
+ */
+@NgModule({
+    declarations: [TranslationPipe, FormatDateTimePipe],
+    exports: [FormatDateTimePipe, TranslationPipe],
+})
+export class I18nModule {
+    /**
+     * Creates a {@link I18nModule} using the global translation service.
+     */
+    static forRoot(): ModuleWithProviders {
+        return {
+            ngModule: I18nModule,
+            providers: [
+                {
+                    provide: TranslationService,
+                    useFactory: genericSingletonFactory,
```

Comment:
I think if you want a single instance, you create it and pass it, why a factory?

Reply:
Because we need access to the client's provided `BOOTSTRAP_DETAILS`, which means that the singleton cannot be constructed until the `BOOTSTRAP_DETAILS` are provided.